### ISSUE TITLE
`mlr --icsv --ojson cat < mlr.bug.csv` drops some columns

### ISSUE THREAD

**Issue by SamuelLarkin** [johnkerl/miller#1523]
Hi,
 I've got a csv file that I wanted to convert to json using `mlr` but columns are disappearing.

Given `mlr.bug.csv`:

```csv
expt_name,baseline-final-test,baseline-final-test.last_checkpoint,stage-all.001,stage-all.001.last_checkpoint
lr=1.0e-5_e=16/000,40.93,39.75,44.05,43.13
lr=1.0e-5_e=16/001,41.09,41.11,43.81,43.82
lr=1.0e-5_e=16/002,40.84,40.81,43.87,43.72
```

If I try the following `mlr --icsv --ojson cat < mlr.bug.csv`, we get:
```json
[
{
  "expt_name": "lr=1.0e-5_e=16/000",
  "baseline-final-test": {
    "last_checkpoint": 39.75
  },
  "stage-all": {
    "001": {
      "last_checkpoint": 43.13
    }
  }
},
{
  "expt_name": "lr=1.0e-5_e=16/001",
  "baseline-final-test": {
    "last_checkpoint": 41.11
  },
  "stage-all": {
    "001": {
      "last_checkpoint": 43.82
    }
  }
},
{
  "expt_name": "lr=1.0e-5_e=16/002",
  "baseline-final-test": {
    "last_checkpoint": 40.81
  },
  "stage-all": {
    "001": {
      "last_checkpoint": 43.72
    }
  }
}
]
```

Note that we've lost `baseline-final-test` probably because it is a substring of `baseline-final-test.last_checkpoint` and same goes for `stage-all.001` which is a substring of `stage-all.001.last_checkpoint`.

An alternate view on the error is that the json creates a new sub-level each time it encounters a `.` in the column's name.

If we try the round trip `mlr --icsv --ojson cat < mlr.bug.csv  | mlr --ijson --ocsv cat` it is also clear that some columns are missing.
```csv
expt_name,baseline-final-test.last_checkpoint,stage-all.001.last_checkpoint
lr=1.0e-5_e=16/000,39.75,43.13
lr=1.0e-5_e=16/001,41.11,43.82
lr=1.0e-5_e=16/002,40.81,43.72
```

# Expectated Output

I was expecting the json output to look something like this assuming I didn't do any cut and paste errors rebuilding the expected output:

```json
[
{
  "expt_name": "lr=1.0e-5_e=16/000",
  "baseline-final-test": 40.93,
  "baseline-final-test.last_checkpoint": 39.75,
  "stage-all.001": 44.05,
  "stage-all.001.last_checkpoint": 43.13,
},
{
  "expt_name": "lr=1.0e-5_e=16/001",
  "baseline-final-test": 41.09,
  "baseline-final-test.last_checkpoint": 41.11,
  "stage-all.001": 43.81,
  "stage-all.001.last_checkpoint": 43.82,
},
{
  "expt_name": "lr=1.0e-5_e=16/002",
  "baseline-final-test": 40.84,
  "baseline-final-test.last_checkpoint": 40.81,
  "stage-all.001": 43.87,
  "stage-all.001.last_checkpoint": 43.72,
}
]
```

**Comment by SamuelLarkin** [comment]
Here's another example

```csv
expt_name,A,B.last_checkpoint,C.001,C.001.last_checkpoint
lr=1.0e-5_e=16/000,40.93,39.75,44.05,43.13
lr=1.0e-5_e=16/001,41.09,41.11,43.81,43.82
lr=1.0e-5_e=16/002,40.84,40.81,43.87,43.72
```

```sh
mlr --icsv --ojson cat < mlr.bug.2.csv
```

```json
[
{
  "expt_name": "lr=1.0e-5_e=16/000",
  "A": 40.93,
  "B": {
    "last_checkpoint": 39.75
  },
  "C": {
    "001": {
      "last_checkpoint": 43.13
    }
  }
},
{
  "expt_name": "lr=1.0e-5_e=16/001",
  "A": 41.09,
  "B": {
    "last_checkpoint": 41.11
  },
  "C": {
    "001": {
      "last_checkpoint": 43.82
    }
  }
},
{
  "expt_name": "lr=1.0e-5_e=16/002",
  "A": 40.84,
  "B": {
    "last_checkpoint": 40.81
  },
  "C": {
    "001": {
      "last_checkpoint": 43.72
    }
  }
}
]
```

**Comment by aborruso** [comment]
Please read this https://miller.readthedocs.io/en/6.11.0/reference-main-flag-list/index.html#flatten-unflatten-flags

The way to do it, is to use `--no-auto-unflatten` flag. Running

```
mlr --c2j --no-auto-unflatten cat input.csv
```

you get

```json
[
  {
    "expt_name": "lr=1.0e-5_e=16/000",
    "baseline-final-test": 40.93,
    "baseline-final-test.last_checkpoint": 39.75,
    "stage-all.001": 44.05,
    "stage-all.001.last_checkpoint": 43.13
  },
  {
    "expt_name": "lr=1.0e-5_e=16/001",
    "baseline-final-test": 41.09,
    "baseline-final-test.last_checkpoint": 41.11,
    "stage-all.001": 43.81,
    "stage-all.001.last_checkpoint": 43.82
  },
  {
    "expt_name": "lr=1.0e-5_e=16/002",
    "baseline-final-test": 40.84,
    "baseline-final-test.last_checkpoint": 40.81,
    "stage-all.001": 43.87,
    "stage-all.001.last_checkpoint": 43.72
  }
]
```

**Comment by SamuelLarkin** [comment]
🤦🏼‍♂️ 
I was going to write that I found a solution but it looks like I'm slightly too late.

Thanks for your answer.

`mlr --icsv --ojson --flatsep : cat < mlr.bug.2.csv`

```json
[
{
  "expt_name": "lr=1.0e-5_e=16/000",
  "A": 40.93,
  "B.last_checkpoint": 39.75,
  "C.001": 44.05,
  "C.001.last_checkpoint": 43.13
},
{
  "expt_name": "lr=1.0e-5_e=16/001",
  "A": 41.09,
  "B.last_checkpoint": 41.11,
  "C.001": 43.81,
  "C.001.last_checkpoint": 43.82
},
{
  "expt_name": "lr=1.0e-5_e=16/002",
  "A": 40.84,
  "B.last_checkpoint": 40.81,
  "C.001": 43.87,
  "C.001.last_checkpoint": 43.72
}
]
```

**Comment by aborruso** [comment]
@SamuelLarkin what do you think about Miller? Have you been using it for a long time?

**Comment by SamuelLarkin** [comment]
@aborruso I think it is great.  I'm fairly new to it.  I regularly have to make csv tables and I came to know `miller` because I needed to prune those tables.  I'm also using it to validate my data using `summary`.  I really like the `then` to have multi-steps processing on the data.  I like the fact that there are a lot of verbs which both a blessing and a curse especially when you are looking for a verb and don't know which one to use.

I've started to spread the good word to my fellow colleagues.

I starred this great project.  I also consider it as part of my essential CLI stack.
